### ISSUE TITLE
RFC: VDiff2: Reimplementing VDiff on tablets

### ISSUE THREAD

**Issue by rohit-nayak-ps** [vitessio/vitess#10134]
# VDiff2: VDiff on tablets

## TL;DR;

VDiff currently uses vtctld. For large tables, VDiff is very slow; cannot be restarted on an error; and can impact app read/write performance. In addition,  vtctld might need to be differently provisioned depending on the size of tables being VDiffed. 

To address these, VDiff 2.0 will run on tablets (one per shard), to address performance and provisioning, and be resumable. 

### Current VDiff Flow

VDiff compares exactly the source and target data at a point in time for a given VReplication workflow. This is a high-level view of the flow.

* Stop target streams
* Start snapshots on the source primaries, record gtids of snapshots
* Start target streams, stopping at the recorded gtids
* Start snapshots on the target primaries
* Stream data from source and target; 
    * Merge-sort the data from different shards as they arrive
    * Make sure every shard has responded
    * Compare the top row. 
    * Rinse and repeat

Note that, since we continue to have writes on the source after the snapshot is taken, the source and target databases *can* diverge. VDiff only guarantees that the data is the same at the time of the snapshot.

#### Single Consistent Snapshot

VDiff uses consistent snapshots and compares the entire table using that snapshot. So it needs to hold that snapshot for the entire duration of the comparison. If there is a network issue then we cannot resume the operation. A fresh snapshot has to be taken and we start VDiff from scratch.

In addition holding the snapshot for so long can impact MVCC (increased history list length) on primaries and replica lags on replicas.

#### Performance

Since a single process is doing the entire comparison, it takes very long. VDiffs have been known to run for several hours or days in some cases. Of course, every time there is a network issue or if a server is restarted/reparented we need to restart it, increasing the perceived latency.

#### Uses vtctld

It runs, not on tablets, but on vtctld. vtctld is generally used as a lightweight wrapper to send commands to vttablets or vtgates. VDiff breaks that abstraction. For large tables it can take a lot of time (hours and days) and require a large amount of memory. This means that, if you want to use vtctld for VDiff the VM needs to be specially provisioned. 

#### Increased network traffic

Also the entire source and target database is pulled by VDiff causing a potentially massive increase in network traffic between tablets and vtctld.

## Proposed Changes

VDiff2 will run on tablets, similar to how VReplication workflows run today. There will be one stream on the primary of each target shard for each VDiff. Each stream will compare the data copied to that target shard from all the source shards that map to it.

VDiff2 is invoked by passing a **-v2** flag to VDiff. The command then schedules one VDiff stream per shard and exits. VDiff2 supports additional actions like **Show/Progress** to view status of a vdiff and **Create/Stop/Start/Restart/Cancel/Resume **for resumable VDiffs.

Since there is no vtctld to orchestrate, similar to online ddl, we need to run one stream per shard. Also, we need to store the results of the VDiff per shard for visibility and resumability. Hence we need to run it off primaries.

Each shard’s vdiff will choose one tablet from each source stream related to the workflow and one for this shard using the tablet picker depending on the cells and tablet types passed to vdiff. If the target is not serving production data, ideally the primary itself should be selected as the source since that tablet is probably closed to the mysql server.

Additionally a VDiff2 can be given a unique name to identify it. In addition it will have a UUID (similar to how Online DDL works today). There will be backing tables in _vt for each shard, containing the progress and persisted state for each vdiff.

By default we will run VDiff2 with a single consistent snapshot and error out if it doesn’t complete. However we will be persisting the last compared PK (once every N rows). So a session can be restarted (or set to automatically restart on error) starting from the lastPK. Note that, in this case, a new consistent snapshot will be taken. If resumability is used, the guarantees provided by VDiff are that the rows compared match as of the snapshot taken when the comparison is performed, not as of a single snapshot.

We will not store the LastPK for every row compared. Updating for every row will result in huge binlogs and we will try to avoid that. 

We already have this problem for vreplication workflows, but for them we need to keep the state consistent with the applied events to ensure correctness. For VDiff it will just mean that some rows will be compared again reducing efficiency without impacting correctness.

### Per-Shard Flow

On every start/resume of a comparison cycle a stream will do the following:
* Stop associated vreplication streams on this target. Record current gtids
* Wait for selected source tablets to reach the gtid 
* Start snapshot on sources, filtered for > lastPK
* Restart target streams to run upto the snapshot gtids
* Start snapshot on target for all data > lastPK
* Restart vreplication streams on target
* Compare the snapshots same as the current VDiff

One change from VDiff will be to use RowStreamer instead of ResultStreamer. RowStreamer allows keyrange filtering and specification of lastpk, which ResultStreamer doesn’t.

### Performance Options

To address performance concerns:
* allow limiting snapshot size  per compare cycle
* Use throttling based on history list length and replica lag

## Further Work

Once the above has been implemented, we plan to add the following features:

### Use checksums

This will allow the source to send only checksums to the target instead of entire rows. This will help reduce the amount of data sent across the network. We could use mysql functions for computing checksums or compute it on the source vttablet. In particular, for huge rows (think blob columns) this will help significantly in simplifying the network/memory requirements.

This will be possible only for exact copies and where the source and target databases have compatible versions (eg. wrt collations).

### Sampling

Instead of the entire database we can just ask VDiff2 to sample, say, 10% of the rows. It will then stream rows from random PK ranges in a table in order to compare these many rows.
